util/ranger/detacher.go

```diff
@@ -86,62 +86,112 @@ func detachColumnDNFConditions(conditions []expression.Expression, checker *cond
 	return accessConditions, hasResidualConditions
 }
 
+// getEqOrInColOffset checks if the expression is a eq function that one side is constant and another is column or an
+// in function which is `column in (constant list)`.
+// If so, it will return the offset of this column in the slice, otherwise return -1 for not found.
+func getEqOrInColOffset(expr expression.Expression, cols []*expression.Column) int {
+	f, ok := expr.(*expression.ScalarFunction)
+	if !ok {
+		return -1
+	}
+	if f.FuncName.L == ast.EQ {
+		if c, ok := f.GetArgs()[0].(*expression.Column); ok {
+			if _, ok := f.GetArgs()[1].(*expression.Constant); ok {
+				for i, col := range cols {
+					if col.Equal(c, nil) {
+						return i
+					}
+				}
+			}
+		}
+		if c, ok := f.GetArgs()[1].(*expression.Column); ok {
```

Comment:
Could we abstract this into a function?

Reply:
We can extract the outer if as a function, but i think it's no need to change this block to function...

Reply:
Both this block and the upper block has the same structure/logic,  so we can extract a function that takes two arguments then call this function twice. This will also reduce the level of nested blocks.